BUG/ENH: Handle NonexistentTimeError in date rounding

mroeschke opened the pull request:
- [x] closes #23324
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Similar strategy as https://github.com/pandas-dev/pandas/pull/22647, added a `nonexistent` keyword argument to `round`, `ceil`, and `floor` to control rounding when encountering a `NonexistentTimeError`

This is also fixes a bug in the `nonexistent='shift'` implementation in https://github.com/pandas-dev/pandas/pull/22644 where dates with timezones with negative UTC offsets got shifted by an additional hour. This bug is naturally tested by these rounding tests

Commit:
nonexistent rounding

Commit:
Fix shifting bug and add tests and whatsnew

pep8speaks commented:
Hello @mroeschke! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/indexes/datetimelike.py`](https://github.com/mroeschke/pandas/blob/35711b08e1b4caddc6639fd96a37ad1ddd6b023d/pandas/core/indexes/datetimelike.py) !

 - There are no PEP8 issues in the file [`pandas/tests/scalar/timestamp/test_unary_ops.py`](https://github.com/mroeschke/pandas/blob/35711b08e1b4caddc6639fd96a37ad1ddd6b023d/pandas/tests/scalar/timestamp/test_unary_ops.py) !

 - There are no PEP8 issues in the file [`pandas/tests/series/test_datetime_values.py`](https://github.com/mroeschke/pandas/blob/35711b08e1b4caddc6639fd96a37ad1ddd6b023d/pandas/tests/series/test_datetime_values.py) !

gfyoung commented on pandas/tests/scalar/timestamp/test_unary_ops.py:
```diff
@@ -155,6 +155,23 @@ def test_round_dst_border(self, method):
+        result = getattr(ts, method)(freq, nonexistent='NaT')
+        assert result is NaT
+
+        with pytest.raises(pytz.NonExistentTimeError):
```
Anything meaningful in the error message?

mroeschke replied:
I think this just raises `NonExistentTimeError: [time that was nonexistent]` but I can test for the message as well.

Commit:
Merge remote-tracking branch 'upstream/master' into nonexistent_round

Commit:
Add expected nonexistent time with raise

gfyoung commented on doc/source/whatsnew/v0.24.0.txt:
```diff
@@ -227,6 +227,7 @@ Other Enhancements
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support an ``ambiguous`` argument for handling datetimes that are rounded to ambiguous times (:issue:`18946`)
+- :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support an ``nonexistent`` argument for handling datetimes that are rounded to nonexistent times (:issue:`22647`)
```
an ``nonexistent`` --> a ``nonexistent``

gfyoung replied:
Do you think this requires a mini-section to explain behavior?

mroeschke replied:
I can link the section in the `timeseries.rst` that explains `nonexistent` behavior.

gfyoung replied:
Awesome.  That works just as well.

Commit:
fix typo and add reference

gfyoung commented on pandas/tests/series/test_datetime_values.py:
```diff
@@ -282,6 +282,27 @@ def test_dt_round_tz_ambiguous(self, method):
+
+        with pytest.raises(pytz.NonExistentTimeError,
+                           message='2018-03-11 02:00:00'):
+            getattr(s.dt, method)(freq, nonexistent='raise')
```
* I'm almost inclined to have parameterization on `nonexistent`, but I'll leave that up to you.
* More general question, to what extent are we using `pytest.raises` vs `tm.assert_raises_regex` ?

cc @jreback

mroeschke replied:
I'm inclined to keep the format of this test. IMO parameterization over nonexistent would obfuscate the test too much.

gfyoung replied:
No problem.  Just a thought. 👍 

More general question, to what extent are we using `pytest.raises` vs `tm.assert_raises_regex` ?

cc @jreback

jreback replied:
ok with ``pytest.raises`` to the extent we are just checking the type of the error. so ok here

gfyoung replied:
@jreback : `pytest.raises` is also checking the error message here...

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -852,6 +851,8 @@ def tz_localize_to_utc(ndarray[int64_t] vals, object tz, object ambiguous=None,
         Py_ssize_t i, idx, pos, ntrans, n = len(vals)
         int64_t *tdata
         int64_t v, left, right, val, v_left, v_right
+        int64_t remaining_minutes, new_local
```
can you add with the previous line here

jreback commented on pandas/_libs/tslibs/conversion.pyx:
```diff
@@ -852,6 +851,8 @@ def tz_localize_to_utc(ndarray[int64_t] vals, object tz, object ambiguous=None,
         int64_t v, left, right, val, v_left, v_right
+        int64_t remaining_minutes, new_local
+        int delta_idx_offset, delta_idx
         ndarray[int64_t] result, result_a, result_b, dst_hours
```
don't we use Py_ssize_t for indexers?

jreback commented on pandas/_libs/tslibs/nattype.pyx:
```diff
@@ -484,6 +484,17 @@ class NaTType(_NaT):
             - 'raise' will raise an AmbiguousTimeError for an ambiguous time
 
             .. versionadded:: 0.24.0
+        nonexistent : 'shift', 'NaT', default 'raise'
```
if any way to share these doc-strings would be great, maybe templating?

mroeschke replied:
@jbrockmendel do you happen to know why for some of these NaT methods we aren't just passing is `Timestamp.[method].__doc__` instead of copying these docstrings?

jbrockmendel replied:
Because of the dependency structure.  We import NaT in timestznps but not the other way around.

jreback commented on pandas/_libs/tslibs/timestamps.pyx:
```diff
@@ -754,14 +756,27 @@ class Timestamp(_Timestamp):
+              existing time
+            - 'NaT' will return NaT where there are nonexistent times
+            - 'raise' will raise an NonExistentTimeError if there are
+              nonexistent times
```
same comment as above about the doc-strings

Commit:
change to py_ssize_t

Commit:
Combine in64_t cdef

mroeschke commented:
Addressed your comments @jreback. I can make another issue between docstring templating between `Timestamp` and `NaT` as many methods share the same docstring (not just `tz_localize`)

jreback commented:
thanks @mroeschke 

yes would like to template the doc-strings to avoid duplication.
